### PR TITLE
Scope all the things

### PR DESCRIPTION
**Pull request by AlCutter**

### CODE REVIEW

File: cpp/client/ct.cc [hunk 1]
```diff
@@ -404,48 +400,46 @@ static void MakeCert() {
...
   // Set the start date to now
-  X509_gmtime_adj(X509_get_notBefore(x), 0);
+  X509_gmtime_adj(X509_get_notBefore(x.get()), 0);
   // End date to now + 1 second
-  X509_gmtime_adj(X509_get_notAfter(x), 1);
+  X509_gmtime_adj(X509_get_notAfter(x.get()), 1);
 
   // Create the issuer name
-  X509_NAME* issuer = X509_NAME_new();
+  X509_NAME* issuer = X509_NAME_new();  // x takes ownership
```
Comment (pphaneuf) [inline cpp/client/ct.cc]: You said you'd use a ScopedSomething here too, and release it at line 414.

File: cpp/client/ct.cc [hunk 2]
```diff
@@ -404,48 +400,46 @@ static void MakeCert() {
...
-  string serialized_sct_list = SCTToList(sct);
-  AddOctetExtension(x, cert_trans::NID_ctSignedCertificateTimestampList,
+  const string serialized_sct_list(SCTToList(sct));
+  AddOctetExtension(x.get(), cert_trans::NID_ctSignedCertificateTimestampList,
                     reinterpret_cast<const unsigned char*>(
                         serialized_sct_list.data()),
                     serialized_sct_list.size(), 1);
 
-  int i = i2d_X509_bio(out, x);
+  int i = i2d_X509_bio(out.get(), x.get());
```
Comment (pphaneuf) [inline cpp/client/ct.cc]: I think that `CHECK_GT` on the next line used to be an `assert`, hence why `i2d_X509_bio` isn't called inline. Could you make it inline while you're there?

File: cpp/client/ssl_client.cc [hunk 3]
```diff
@@ -65,29 +65,29 @@ int SSLClient::ExtensionCallback(SSL*, unsigned ext_type,
...
     : client_(server, port),
-      ctx_(NULL),
-      ssl_(NULL),
+      ctx_(SSL_CTX_new(TLSv1_client_method())),
+      ssl_(),
       verify_args_(verifier),
       connected_(false) {
-  ctx_ = SSL_CTX_new(TLSv1_client_method());
-  CHECK_NOTNULL(ctx_);
+  CHECK_NOTNULL(ctx_.get());
```
Comment (pphaneuf) [inline cpp/client/ssl_client.cc]: Should put that inline at line 68.

File: cpp/client/ssl_client.cc [hunk 4]
```diff
@@ -96,8 +96,6 @@ SSLClient::SSLClient(const string& server, uint16_t port, const string& ca_dir,
 
 SSLClient::~SSLClient() {
   Disconnect();
-  if (ctx_ != NULL)
-    SSL_CTX_free(ctx_);
```
Comment (pphaneuf) [inline cpp/client/ssl_client.cc]: Would you want to make `verify_args_.verifier` a `unique_ptr`, while you're at it?

File: cpp/client/ssl_client.cc [hunk 5]
```diff
@@ -106,10 +104,8 @@ bool SSLClient::Connected() const {
 }
 
 void SSLClient::Disconnect() {
-  if (ssl_ != NULL) {
-    SSL_shutdown(ssl_);
-    SSL_free(ssl_);
-    ssl_ = NULL;
```
Comment (pphaneuf) [inline cpp/client/ssl_client.cc]: You need some `ssl_.reset()` here!

File: cpp/client/ssl_client.cc [hunk 6]
```diff
@@ -279,15 +275,15 @@ SSLClient::HandshakeResult SSLClient::SSLConnect(bool strict) {
...
-  CHECK_NOTNULL(ssl_);
-  BIO* bio = BIO_new_socket(client_.fd(), BIO_NOCLOSE);
-  CHECK_NOTNULL(bio);
+  ssl_.reset(SSL_new(ctx_.get()));
+  CHECK_NOTNULL(ssl_.get());
+  ScopedBIO bio(BIO_new_socket(client_.fd(), BIO_NOCLOSE));
+  CHECK_NOTNULL(bio.get());
   // Takes ownership of bio.
-  SSL_set_bio(ssl_, bio, bio);
+  SSL_set_bio(ssl_.get(), bio.get(), bio.release());
```
Comment (pphaneuf) [inline cpp/client/ssl_client.cc]: Can't do that: `bio.release()` could be called before `bio.get()`. :boom:

File: cpp/log/cert.cc [hunk 7]
```diff
@@ -380,15 +358,14 @@ StatusOr<bool> Cert::IsSignedBy(const Cert& issuer) const {
     return util::Status(Code::FAILED_PRECONDITION, "Cert not loaded");
   }
 
-  EVP_PKEY* const issuer_key = X509_get_pubkey(issuer.x509_);
-  if (issuer_key == NULL) {
+  ScopedEVP_PKEY issuer_key(X509_get_pubkey(issuer.x509_.get()));
```
Comment (pphaneuf) [inline cpp/log/cert.cc]: Make it `const` too?

File: cpp/log/cert.h [hunk 8]
```diff
@@ -27,12 +28,11 @@ class Cert {
...
   // All attempts to operate on an unloaded cert will fail with ERROR.
   explicit Cert(const std::string& pem_string);
-  Cert() : x509_(NULL) {
+  Cert() : x509_() {
   }
-  ~Cert();
 
   bool IsLoaded() const {
-    return x509_ != NULL;
+    return x509_ != nullptr;
```
Comment (pphaneuf) [inline cpp/log/cert.h]: `return x509_;`!

File: cpp/log/cms_verifier.cc [hunk 9]
```diff
@@ -71,9 +70,9 @@ StatusOr<bool> CmsVerifier::IsCmsSignedByCert(const string& cms_object,
   // Now that we've got the CMS unpacked check it has a valid signature using
   // the same key as the cert. First create a certificate stack from our
   // expected signing cert that can be used by CMS_verify.
-  STACK_OF(X509)* validation_chain = sk_X509_new(nullptr);
+  ScopedX509Stack validation_chain(sk_X509_new(nullptr));
 
-  sk_X509_push(validation_chain, cert->x509_);
+  sk_X509_push(validation_chain.get(), cert->x509_.get());
```
Comment (pphaneuf) [inline cpp/log/cms_verifier.cc]: Maybe `release()`? Or something else?

File: cpp/log/signer.cc [hunk 10]
```diff
@@ -50,7 +46,7 @@ void Signer::Sign(const std::string& data,
 }
 
 Signer::Signer()
-    : pkey_(NULL),
+    : pkey_(),
```
Comment (pphaneuf) [inline cpp/log/signer.cc]: Could just ding all those default constructor invocations.

File: cpp/client/ssl_client.cc [hunk 11]
```diff
@@ -65,29 +65,29 @@ int SSLClient::ExtensionCallback(SSL*, unsigned ext_type,
 SSLClient::SSLClient(const string& server, uint16_t port, const string& ca_dir,
                      LogVerifier* verifier)
     : client_(server, port),
-      ctx_(NULL),
-      ssl_(NULL),
+      ctx_(SSL_CTX_new(TLSv1_client_method())),
+      ssl_(),
```
Comment (pphaneuf) [inline cpp/client/ssl_client.cc]: Ding default constructor invocation.

File: cpp/log/verifier.cc [hunk 12]
```diff
@@ -69,7 +65,7 @@ std::string Verifier::ComputeKeyID(EVP_PKEY* pkey) {
 }
 
 Verifier::Verifier()
-    : pkey_(NULL),
+    : pkey_(),
```
Comment (pphaneuf) [inline cpp/log/verifier.cc]: Ding default constructor invocation.

File: cpp/log/cert.cc [hunk 13]
```diff
@@ -77,17 +77,16 @@ Cert::Cert(X509* x509) : x509_(x509) {
 }
 
 
-Cert::Cert(const string& pem_string) : x509_(nullptr) {
+Cert::Cert(const string& pem_string) : x509_() {
```
Comment (pphaneuf) [inline cpp/log/cert.cc]: Ding default constructor invocation.

File: cpp/log/cert.cc [hunk 14]
```diff
@@ -1102,25 +1057,19 @@ util::Status Cert::IsValidNameConstrainedIntermediateCa() const {
 }
 
 
-TbsCertificate::TbsCertificate(const Cert& cert) : x509_(nullptr) {
+TbsCertificate::TbsCertificate(const Cert& cert) : x509_() {
```
Comment (pphaneuf) [inline cpp/log/cert.cc]: Ding default constructor invocation.

File: cpp/log/cert.h [hunk 15]
```diff
@@ -27,12 +28,11 @@ class Cert {
   // so caller should check IsLoaded() before doing anything else.
   // All attempts to operate on an unloaded cert will fail with ERROR.
   explicit Cert(const std::string& pem_string);
-  Cert() : x509_(NULL) {
+  Cert() : x509_() {
```
Comment (pphaneuf) [inline cpp/log/cert.h]: Ding default constructor invocation (or the whole thing, really).

File: cpp/client/ct.cc (hunk 1 above)
Comment (AlCutter) [inline cpp/client/ct.cc]: Yeah, this one got away.
Done.

File: cpp/client/ct.cc (hunk 2 above)
Comment (AlCutter) [inline cpp/client/ct.cc]: Done

File: cpp/client/ssl_client.cc (hunk 3 above)
Comment (AlCutter) [inline cpp/client/ssl_client.cc]: Done

File: cpp/client/ssl_client.cc (hunk 4 above)
Comment (AlCutter) [inline cpp/client/ssl_client.cc]: Done

File: cpp/client/ssl_client.cc (hunk 5 above)
Comment (AlCutter) [inline cpp/client/ssl_client.cc]: Good spot, got a bit overzealous on the delete!

File: cpp/client/ssl_client.cc (hunk 6 above)
Comment (AlCutter) [inline cpp/client/ssl_client.cc]: Good catch, fixed.

File: cpp/log/cert.cc (hunk 7 above)
Comment (AlCutter) [inline cpp/log/cert.cc]: Done

File: cpp/log/cert.h (hunk 8 above)
Comment (AlCutter) [inline cpp/log/cert.h]: No, doesn't work - I tried that initially.

```
./cpp/log/cert.h:35:12: error: no viable conversion from 'const ScopedX509' (aka 'const std::unique_ptr<x509_st, OpenSSLDeleter<x509_st,
      &X509_free> >') to 'bool'
    return x509_;
```

File: cpp/log/cms_verifier.cc (hunk 9 above)
Comment (AlCutter) [inline cpp/log/cms_verifier.cc]: Something else I think; Looks like the stack should be a `weak` one.

File: cpp/log/signer.cc (hunk 10 above)
Comment (AlCutter) [inline cpp/log/signer.cc]: Yeah, I was going for minimal changes, but it's got past that now...
Dinged.

File: cpp/client/ssl_client.cc (hunk 11 above)
Comment (AlCutter) [inline cpp/client/ssl_client.cc]: Dinged.

File: cpp/log/verifier.cc (hunk 12 above)
Comment (AlCutter) [inline cpp/log/verifier.cc]: Dinged

File: cpp/log/cert.cc (hunk 13 above)
Comment (AlCutter) [inline cpp/log/cert.cc]: Dinged

File: cpp/log/cert.cc (hunk 14 above)
Comment (AlCutter) [inline cpp/log/cert.cc]: Dinged

File: cpp/log/cert.h (hunk 15 above)
Comment (AlCutter) [inline cpp/log/cert.h]: Dinged (not the whole thing)

File: cpp/log/cert.h (hunk 8 above)
Comment (pphaneuf) [inline cpp/log/cert.h]: Bleh, there's an `operator bool()`, but it's marked explicit. The compiler could tell you that! :stuck_out_tongue: 

I've used `.get()` for this before, but this is fine, if you're going to have a bit of fluff, one or the other, doesn't matter much...

File: cpp/client/ssl_client.cc [hunk 16]
```diff
@@ -154,7 +147,7 @@ LogVerifier::LogVerifyResult SSLClient::VerifySCT(const string& token,
 int SSLClient::VerifyCallback(X509_STORE_CTX* ctx, void* arg) {
   VerifyCallbackArgs* args = reinterpret_cast<VerifyCallbackArgs*>(arg);
   CHECK_NOTNULL(args);
-  LogVerifier* verifier = args->verifier;
+  LogVerifier* verifier(args->verifier.get());
```
Comment (pphaneuf) [inline cpp/client/ssl_client.cc]: That variable is a bit of a waste of time, no? :smiley: 

(don't have to fix it, your call)

**Comment by pphaneuf** [comment]
LGTM

**Pull request merged**